File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expand(l, r):
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            return r - l - 1
        res = [0, 0]
        for i in range(len(s)-1):
            odd_len = expand(i, i)
            if res[1] - res[0] + 1 < odd_len:
                dist = odd_len // 2
                res = [i - dist, i + dist]

            even_len = expand(i, i+1)
            if res[1] - res[0] + 1 < even_len:
                dist = even_len // 2 - 1
                res = [i - dist, i + dist + 1]
        return s[res[0] : res[1] + 1]

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int[] res = new int[2];
        for (int i = 0; i < s.length(); i++) {
            int oddLen = expand(s, i, i);
            if (oddLen > res[1] - res[0] + 1) {
                int dist = oddLen / 2;
                res[0] = i - dist;
                res[1] = i + dist;
            }
            int evenLen = expand(s, i, i+1);
            if (evenLen > res[1] - res[0] + 1) {
                int dist = evenLen / 2 - 1;
                res[0] = i - dist;
                res[1] = i + 1 + dist;
            }
        }
        return s.substring(res[0], res[1]+1);
    }
    private int expand(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }
}